opm/simulators/wells/ParallelWellInfo.hpp

```diff
@@ -79,13 +82,88 @@ class CommunicateAbove
     /// \param current C-array of the values at the perforations
     /// \param size The size of the C-array and the returned vector
     /// \return a vector containing the values for the perforation above.
-    std::vector<double> communicate(double first_value,
-                                    const double* current,
-                                    std::size_t size);
+    std::vector<double> communicateAbove(double first_value,
+                                         const double* current,
+                                         std::size_t size);
 
-private:
+    /// \brief Creates an array of values for the perforation below.
+    /// \param first_value Value to use for above of the first perforation
+    /// \param current C-array of the values at the perforations
+    /// \param size The size of the C-array and the returned vector
+    /// \return a vector containing the values for the perforation above.
+    std::vector<double> communicateBelow(double first_value,
+                                         const double* current,
+                                         std::size_t size);
+
+    /// \brief Do a (in place) partial sum on values attached to all perforations.
+    ///
+    /// For distributed wells this may include perforations stored elsewhere.
+    /// The result is stored in ther range given as the parameters
+    /// \param begin The start of the range
+    /// \param ebd The end of the range
+    /// \tparam RAIterator The type og random access iterator
+    template<class RAIterator>
+    void partialSumPerfValues(RAIterator begin, RAIterator end) const
+    {
+        if (this->comm_.size() < 2)
+        {
+            std::partial_sum(begin, end, begin);
+        }
+        else
+        {
 #if HAVE_MPI
+            // The global index used in the index set current_indices
+            // is the index of the perforation in ECL Schedule definition.
+            // This is assumed to give the topological order that is used
+            // when doing the partial sum.
+            // allgather the index of the perforation in ECL schedule and the value.
+            using Value = typename std::iterator_traits<RAIterator>::value_type;
+            std::vector<int> sizes(comm_.size());
+            std::vector<int> displ(comm_.size(), 0);
+            using GlobalIndex = typename IndexSet::IndexPair::GlobalIndex;
+            using Pair = std::pair<GlobalIndex,Value>;
+            std::vector<Pair> my_pairs;
+            my_pairs.reserve(current_indices_.size());
+            for (const auto& pair: current_indices_)
+            {
+                if (pair.local().attribute() == owner)
+                {
+                    my_pairs.emplace_back(pair.global(), *(begin+(static_cast<const int&>(pair.local()))));
```

Comment:
Looks correct, but why not just `begin[pair.local()]`?

Also on line 156.

Reply:
I have no idea why I did not do that. Done now. Will merge once test has run.